spec/controllers/checkout_controller_spec.rb

```diff
@@ -266,4 +266,80 @@
       controller.send(:update_failed)
     end
   end
+
+  # This is the first example of testing concurrency in the Open Food Network.
+  # If we want to do this more often, we should look at:
+  #
+  #   https://github.com/forkbreak/fork_break
+  #
+  # The concurrency flag enables multiple threads to see the same database
+  # without isolated transactions.
+  describe "concurrency", concurrency: true do
+    # Re-defining test data because we need full data without doubles.
+    # :order_cycle has suppliers, distributors and products.
+    let(:order_cycle) { create(:order_cycle) }
+    let(:distributor) { order_cycle.distributors.first }
+    let(:order) { create(:order, order_cycle: order_cycle, distributor: distributor) }
+    let(:address) { create(:address) }
+    let(:payment_method) { create(:payment_method, distributors: [distributor]) }
+    let(:breakpoint) { Mutex.new }
+
+    before do
+      # Create a valid order ready for checkout:
+      create(:shipping_method, distributors: [distributor])
+      variant = order_cycle.variants_distributed_by(distributor).first
+      order.line_items << create(:line_item, variant: variant)
+
+      # Set up controller environment:
+      session[:order_id] = order.id
+      allow(controller).to receive(:spree_current_user).and_return(order.user)
+      allow(controller).to receive(:current_distributor).and_return(order.distributor)
+      allow(controller).to receive(:current_order_cycle).and_return(order.order_cycle)
+
+      # New threads start running straight away. The breakpoint is after loading
+      # the order and before advancing the order's state and making payments.
+      breakpoint.lock
+      allow(controller).to receive(:check_order_for_phantom_fees) do
+        breakpoint.synchronize {}
+      end
+    end
+
+    it "waits for concurrent checkouts" do
+      # Basic data the user submits during checkout:
+      address_params = address.attributes.except("id")
+      order_params = {
+        "payments_attributes" => [
+          {
+            "payment_method_id" => payment_method.id,
+            "amount" => order.total
+          }
+        ],
+        "bill_address_attributes" => address_params,
+        "ship_address_attributes" => address_params,
+      }
+
+      # Starting two checkout threads. The first one will wait at the breakpoint.
```

Comment:
I am curious about this: the breakpoint lock and synchronize instructions above will only be taken by the first thread? is that what this means?

Reply:
Yes. If we didn't have the around filter then both threads would wait at the breakpoint. But since I implemented the locking variants in the around filter, the second thread will wait there and not get the breakpoint until the first thread is finished.

Reply:
ah, I get it, thanks for explaining!